### DISCUSSION TITLE
Achordion and Tap Dance

### DISCUSSION THREAD

**Discussion by drashna** [getreuer/qmk-keymap#45] [General]
I'm not using it personally, but I was wondering if it would be possible to get achordion working with the quad tap dance functionality?   Eg:
https://docs.qmk.fm/#/feature_tap_dance?id=example-4

**Comment by getreuer** [comment]
Hi @drashna, thanks for pointing this out! I didn't realize that Achordion has a conflict with Tap Dance. The intention is that Achordion only modifies `MT` and `LT` hold events, possibly revising holds to taps. Achordion's event handler tests whether the current key is a tap-hold key with this snippet:

https://github.com/getreuer/qmk-keymap/blob/main/features/achordion.c#L83C14-L84C62

```c
// Determine whether the current event is for a mod-tap or layer-tap key.
const bool is_mt = IS_QK_MOD_TAP(keycode);
const bool is_tap_hold = is_mt || IS_QK_LAYER_TAP(keycode);
```

Then if the key is a Tap Dance or another non-tap-hold key, the intention is the event passes along without change:

* If Achordion's state is "released" or "settled," the event passes directly (`return true` from the handler).
* If Achordion is in "unsettled" state, then (1) Tap Dance event is copied and blocked (`return false), (2) the unsettled tap-hold key is settled, and (3) the Tap Dance event is sent again by passing it to `process_record()`.

The first "released" or "settled" case is straightforward. I'd expect Tap Dance and other non-tap-hold keys to function as expected when used in isolation.

As we've discussed before regarding Auto Mouse Layer, Achordion's handling of the second "unsettled" case is problematic for some features, which may end up receiving the event twice. If this is the problem, it occurs when using a `MT` or `LT` key at the same time as the Tap Dance key.

Is this consistent with what you are observing? Let me know what I'm missing.

**Comment by drashna** [comment]
Yeah, basically.  It doesn't recognize the tap dance as a tap hold key.  The key functions normally, but with none of the achordion-ness. 

And I suspect it's not as simple as just adding a "is tap dance" check here. 


As for the mouse layer thing, the only real method I've seen of fixing that is: 
<https://github.com/drashna/qmk_firmware/commit/0e55b824a637e92b61f2dbda24fbbaf5969666b0>
and
<https://github.com/drashna/qmk_userspace/commit/072a8480b1178cb0fb35ba70781c2cb8ca5441c8>

Eg, reading the tracker value, saving it, and then setting it afterwards.  It's a crap workaround, but it seems to work well.
